### ISSUE TITLE
warning, low memory and too high rfo-basic memory usage

### ISSUE THREAD

**Issue by JanuszChmiel** [RFO-BASIC/Basic#226]
Dear developers,
Rfo-basic is very powerful tool. Basic is modern, support structured programming. Thanks to Nicolas, there is more and more powerful tools, including profi GW library. But when I use rfo-basic more and more often, I see one big problem. Too high memory allocations. A specially if program code runs by using do
until
I have found out, that removing code which is responsible for displaiing low memory is not effective way. Because sooner or later, Android operating system display Warning low memory. The issue is, that Basic allocates memory itself while running in cycles. So here is no way for programmer to solve that. I Am afraid, that using garbage collector to reduce The memory footprint by recalling this Android build in tool is not effective process.
The very big issue is, how to eliminate so big memory allocations. The most difficult problem in this area is, that I think, that Java code of run.java is written effectively. All variables are cleared when possible, all methods are calling to release The memory which have been used by some methods.
I have opened this discussion issue to open development debate how to fix this.
I have used several apps which have been written completely in Java including Timer, and I have never read Warning low memory from Android. So please, do not think about me, that I want to make you sad. I Am aware, that Basic implementation by using Java language is very complex. But I really think, that there should be some solution to solve memory allocations.
Sure, if programmer will try to work with much amount of data, it is normal.
But if some cycle will only count one numeric variable by counting A =a +1, developer is relatively helpless to solve memory leaks which have been caused by Rfo-Basic interpreter itself.
I know, that there is very probably not simple solution for that. Yes, using Nexus 6 or newest Samsung mobile phone, which have 2 GB of RAM.
May be, that Android newer than 4.1.2 is much more capable to handle memory than my old 4.1.2 used Android version.
Thank you very much and I hope, that I did not make somebody to feel sad.
Some article which could help.

https://developer.android.com/training/articles/perf-tips.html

**Comment by olimaticer** [comment]
Hi Janusz,
  The most reasons for memory faults:
1. Loading bitmaps and audio files again and again, but not deleting or releasing them after use.
2. The try to load bitmaps larger than about double device screen size.

BASIC! use only numeric variables from type Double.
That is a fact and today not negotiable.
In future may be Float values, too.
But thas solves not mainly a memory problem.

In difference to Java, BASIC! has not a TRY function.
But if you study the Java source code of BASIC!, you see the TRY command is often used to handle critical steps.

I see you use also 
**OnLowMemory:
LOWMEMORY.RESUME**
That are commands to handle this cases, but not really a substitute for a TRY command.

Janusz, please explain, why A =a +1 in a loop is a problem.
OK, when you cross the limits of a value from type Double.
But that is a programmers problem in all languages.
Why should BASIC! request more memory at overwriting a given value?


Gregor

**Comment by JanuszChmiel** [comment]
Dear Oliver,
I Am very sorry for my bad opinion related to loop.
I have Samsung Galaxy chat GBb-5330, with 512 MB of RAM, but only 434 MB 
of RAm can be used. And because I Am using screen reader and Espeak, 
there are other 10S of MB allocated while using it. And before I have 
started to experiment with low memory warning subpression steps, I have 
run Basic and I have done no actions.
For example, I have only typed
Input "enter text", t$
But after some 1 or 1 minute 30 seconds, Rfo-basic generated low memory 
warning. So I have though, that Basic automatically allocates more and 
more RAM. It is true, that my second mobile phone, which have 1 GB of 
RAM and uses newer version of Android do not have this problem.
I have 430 MB of free rAM, especially because manufacturer of my second 
phone did not include much extra of software. Only clean Android and 
Google opensource apps.
When I will use new audio.load command, which support live Internet streams.
How should I release The loaded stream? by using audio.stop
I do not know, if audio.stop erase audio table index and allocated memory.
Thank You that you have explained me so much information about memory 
allocations in Rfo-basic.
I will also close this issue, because you have explained it and it was 
also not good idea that I have opened it. I Am very sorry for that.
And thank you very much for yours support.
